File: tests/spring/catalog/src/main/java/com/example/catalog/service/AuditService.java
package com.example.catalog.service;

import com.example.catalog.domain.AuditEntry;
import com.example.catalog.mapper.AuditMapper;
import java.time.Clock;
import java.time.LocalDateTime;
import org.jspecify.annotations.Nullable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Propagation;
import org.springframework.transaction.annotation.Transactional;

/** 監査ログ。呼び出し元がロールバックしても残るように、別のトランザクションで書く。 */
@Service
public class AuditService {

    private final AuditMapper auditMapper;
    private final Clock clock;

    public AuditService(AuditMapper auditMapper, Clock clock) {
        this.auditMapper = auditMapper;
        this.clock = clock;
    }

    @Transactional(propagation = Propagation.REQUIRES_NEW)
    public void record(String action, @Nullable Long itemId) {
        AuditEntry entry = new AuditEntry();
        entry.setAction(action);
        entry.setItemId(itemId);
        entry.setCreatedAt(LocalDateTime.now(clock));
        auditMapper.insert(entry);
    }

    @Transactional(readOnly = true)
    public long count() {
        return auditMapper.count();
    }
}

File: tests/spring/catalog/expected/src/service/audit_service.rs
//! Translated from `AuditService` (AuditService.java) by Java2RustTranslator.

use std::sync::Arc;

use sqlx::{PgConnection, PgPool};

use crate::clock::Clock;
use crate::domain::AuditEntry;
use crate::error::{Error, Result};
use crate::mapper::audit_mapper;

/// 監査ログ。呼び出し元がロールバックしても残るように、別のトランザクションで書く。
#[derive(Clone)]
pub struct AuditService {
    pool: PgPool,
    clock: Arc<dyn Clock>,
}

impl AuditService {
    pub fn new(pool: PgPool, clock: Arc<dyn Clock>) -> Self {
        Self { pool, clock }
    }

    /// `@Transactional(propagation = REQUIRES_NEW)`
    pub async fn record(&self, action: &str, item_id: Option<i64>) -> Result<()> {
        let mut tx = self.pool.begin().await?;
        let result = self.record_in(&mut tx, action, item_id).await;
        if matches!(result, Ok(_) | Err(Error::SoldOut(..))) {
            tx.commit().await?;
        }
        result
    }

    /// `record` の本体。呼び出し元の接続（トランザクション）で実行する。
    pub(crate) async fn record_in(
        &self,
        conn: &mut PgConnection,
        action: &str,
        item_id: Option<i64>,
    ) -> Result<()> {
        let entry = AuditEntry {
            action: action.to_string(),
            item_id,
            created_at: self.clock.now(),
        };
        audit_mapper::insert(conn, &entry).await?;
        Ok(())
    }

    /// `@Transactional(readOnly = true)`
    pub async fn count(&self) -> Result<i64> {
        let mut tx = self.pool.begin().await?;
        let result = self.count_in(&mut tx).await;
        if matches!(result, Ok(_) | Err(Error::SoldOut(..))) {
            tx.commit().await?;
        }
        result
    }

    /// `count` の本体。呼び出し元の接続（トランザクション）で実行する。
    async fn count_in(&self, conn: &mut PgConnection) -> Result<i64> {
        Ok(audit_mapper::count(conn).await?)
    }
}
